__________________________________________________________________________________________________________________________________________________________________

# Prediction of arrest for crimes committed in Chicago, Illinois, USA.

__________________________________________________________________________________________________________________________________________________________________

# Table of Contents
- [Introduction](#Introduction)
- [Key Question](#Key-Question)
- [Methods](#Methods)
- [Feature Overview](#Feature-Overview)
- [Loading and Setup](#Loading-and-Setup)
- [Assessment](#Assessment)
- [Targeted Data Frame](#Targeted-Data-Frame)
- [Irrelevant/Redundant Features ](#Irrelevant-/-Redundant-Features)
- [Checking for Duplicate Data](#Checking-for-Duplicate-Data)
- [Dealing with Missing Data](#Dealing-with-Missing-Data)
- [Saving the Data](#Saving-the-Data)

__________________________________________________________________________________________________________________________________________________________________

## Introduction <a id='Introduction'></a>

This project focuses on constructing a machine learning model designed to predict future arrests for crimes committed in Chicago accurately. The potential of this model lies in its ability to aid the Chicago Police Department in curbing crime rates within the city.

__________________________________________________________________________________________________________________________________________________________________

## Key Question <a id='Key-Question'></a>

### The analysis aims to answer this one question:

Using machine learning, how might we accurately predict the future arrest for committed crimes.
This inquiry forms a crucial basis for minimizing crime in Chicago. The dataset utilized originates from the Chicago Data Portal https://data.cityofchicago.org/


__________________________________________________________________________________________________________________________________________________________________

## Methods <a id='Methods'></a>
I used the Python programming language along with libraries like pandas for data manipulation, matplotlib for visualizing data, and sklearn for machine learning and data preprocessing tasks. My choice  of integrated development environment was Jupyter notebook, providing an interactive platform for programming and visualization.

__________________________________________________________________________________________________________________________________________________________________

## Feature Overview <a id='Feature-Overview'></a>

- **ID**                            Unique identifier for the record
- **Case Number**                   The Chicago Police Department RD Number (Records Division Number), which is unique to the incident.
- **Date**                          Date when the incident occurred
- **Block**                         The partially redacted address where the incident occurred, placing it on the same block as the actual address
- **IUCR**                          The Illinois Unifrom Crime Reporting code
- **Primary Type**                  The primary description of the IUCR code
- **Location Description**          Description of the location where the incident occurred 
- **Arrest**                        Indicates whether an arrest was made
- **Domestic**                      Indicates whether the incident was domestic-related as defined by the Illinois Domestic Violence Act.
- **Beat**                          Indicates the beat where the incident occurred. A beat is the smallest police geographic area – 
- **District**                      Indicates the police district where the incident occurred
- **Ward**                          The ward (City Council district) where the incident occurred
- **Community Area**                Indicates the community area where the incident occurred.
- **FBI Code**                      Indicates the crime classification as outlined in the FBI's National Incident-Based Reporting System (NIBRS)
- **X Coordinate**                  The x coordinate of the location where the incident occurred
- **Y Coordinate**                  The y coordinate of the location where the incident occurred
- **Year**                          Year the incident occurred
- **Updated On**                    Date and time the record was last updated 
- **Latitude**                      The latitude of the location where the incident occurred
- **Longitude**                     The longitude of the location where the incident occurred
- **Location**                      The location where the incident occurred in a format that allows for creation of maps and other geographic operations on this                                     data portal.

## Loading and Setup <a id='Loading-and-Setup'></a>

In [49]:
# import relevant packages
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import statsmodels.api as sm
from scipy import stats
from scipy.stats import norm

In [50]:
# reading file 
initial_df = pd.read_csv('/Users/vikst/Desktop/data/Crimes_-_2001_to_Present.csv')

In [51]:
# sanity check
initial_df.head()

ID Case Number                    Date                 Block  IUCR  \
0   5741943    HN549294  08/25/2007 09:22:18 AM    074XX N ROGERS AVE  0560   
1   1930689    HH109118  01/05/2002 09:24:00 PM        007XX E 103 ST  0820   
2  13203321    JG415333  09/06/2023 05:00:00 PM      002XX N Wells st  1320   
3  13210088    JG423627  08/31/2023 12:00:00 PM  023XX W JACKSON BLVD  1153   
4  13210004    JG422532  07/24/2023 09:45:00 PM  073XX S JEFFERY BLVD  0281   

              Primary Type                          Description  \
0                  ASSAULT                               SIMPLE   
1                    THEFT                       $500 AND UNDER   
2          CRIMINAL DAMAGE                           TO VEHICLE   
3       DECEPTIVE PRACTICE  FINANCIAL IDENTITY THEFT OVER $ 300   
4  CRIMINAL SEXUAL ASSAULT                       NON-AGGRAVATED   

                     Location Description  Arrest  Domestic  ...  Ward  \
0                                   OTHER   False     False  ...  49.0   
1                             GAS STATION    True     False  ...   NaN   
2  PARKING LOT / GARAGE (NON RESIDENTIAL)   False     False  ...  42.0   
3                                  STREET   False     False  ...  27.0   
4                               APARTMENT   False     False  ...   7.0   

   Community Area  FBI Code  X Coordinate Y Coordinate  Year  \
0             1.0       08A           NaN          NaN  2007   
1             NaN        06           NaN          NaN  2002   
2            32.0        14           NaN          NaN  2023   
3            28.0        11     1160870.0    1898642.0  2023   
4            43.0        02     1190812.0    1856743.0  2023   

               Updated On   Latitude  Longitude                      Location  
0  08/17/2015 03:03:40 PM        NaN        NaN                           NaN  
1  02/04/2016 06:33:39 AM        NaN        NaN                           NaN  
2  09/14/2023 03:43:09 PM        NaN        NaN                           NaN  
3  09/16/2023 03:41:56 PM  41.877565 -87.684791  (41.877565108, -87.68479102)  
4  09/16/2023 03:41:56 PM  41.761919 -87.576209   (41.7619185, -87.576209245)  

[5 rows x 22 columns]

## Assessment <a id='Assessment'></a>

In [52]:
#gitting shape of data  
initial_df.shape

(7898456, 22)

In [53]:
print("The original data has {} rows". format(initial_df.shape[0]))
print("The original data has {} columns".format(initial_df.shape[1]))

The original data has 7898456 rows
The original data has 22 columns


In [54]:
## Quickly check the data types
initial_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7898456 entries, 0 to 7898455
Data columns (total 22 columns):
 #   Column                Dtype  
---  ------                -----  
 0   ID                    int64  
 1   Case Number           object 
 2   Date                  object 
 3   Block                 object 
 4   IUCR                  object 
 5   Primary Type          object 
 6   Description           object 
 7   Location Description  object 
 8   Arrest                bool   
 9   Domestic              bool   
 10  Beat                  int64  
 11  District              float64
 12  Ward                  float64
 13  Community Area        float64
 14  FBI Code              object 
 15  X Coordinate          float64
 16  Y Coordinate          float64
 17  Year                  int64  
 18  Updated On            object 
 19  Latitude              float64
 20  Longitude             float64
 21  Location              object 
dtypes: bool(2), float64(7), int64(3), object(1

In [55]:
# Getting genearl inderstanding of number of crimes per year
initial_df['Year'].value_counts()

2002    486807
2001    485888
2003    475985
2004    469423
2005    453774
2006    448178
2007    437090
2008    427184
2009    392827
2010    370517
2011    351993
2012    336321
2013    307540
2014    275792
2016    269834
2017    269109
2018    268912
2015    264793
2019    261352
2022    238945
2020    212218
2021    208870
2023    185104
Name: Year, dtype: int64

## Targeted data frame <a id ='Targeted-Data-Frame'></a>

Creating a new dataframe, and getting information only from 2021 to present.

In [56]:
df = initial_df[initial_df['Year']>=2021]
# Reseting index
df.reset_index(drop=True, inplace=True)

In [57]:
df.head()

ID Case Number                    Date                 Block  IUCR  \
0  13203321    JG415333  09/06/2023 05:00:00 PM      002XX N Wells st  1320   
1  13210088    JG423627  08/31/2023 12:00:00 PM  023XX W JACKSON BLVD  1153   
2  13210004    JG422532  07/24/2023 09:45:00 PM  073XX S JEFFERY BLVD  0281   
3  13210062    JG423596  08/27/2023 07:00:00 AM  034XX N LAWNDALE AVE  0820   
4  13210107    JG411849  09/04/2023 09:30:00 PM     053XX S HOMAN AVE  1310   

              Primary Type                          Description  \
0          CRIMINAL DAMAGE                           TO VEHICLE   
1       DECEPTIVE PRACTICE  FINANCIAL IDENTITY THEFT OVER $ 300   
2  CRIMINAL SEXUAL ASSAULT                       NON-AGGRAVATED   
3                    THEFT                       $500 AND UNDER   
4          CRIMINAL DAMAGE                          TO PROPERTY   

                     Location Description  Arrest  Domestic  ...  Ward  \
0  PARKING LOT / GARAGE (NON RESIDENTIAL)   False     False  ...  42.0   
1                                  STREET   False     False  ...  27.0   
2                               APARTMENT   False     False  ...   7.0   
3                               APARTMENT   False     False  ...  30.0   
4                      RESIDENCE - GARAGE   False     False  ...  14.0   

   Community Area  FBI Code  X Coordinate Y Coordinate  Year  \
0            32.0        14           NaN          NaN  2023   
1            28.0        11     1160870.0    1898642.0  2023   
2            43.0        02     1190812.0    1856743.0  2023   
3            21.0        06     1151117.0    1922554.0  2023   
4            63.0        14     1154617.0    1869046.0  2023   

               Updated On   Latitude  Longitude                       Location  
0  09/14/2023 03:43:09 PM        NaN        NaN                            NaN  
1  09/16/2023 03:41:56 PM  41.877565 -87.684791   (41.877565108, -87.68479102)  
2  09/16/2023 03:41:56 PM  41.761919 -87.576209    (41.7619185, -87.576209245)  
3  09/16/2023 03:41:56 PM  41.943379 -87.719974    (41.943378528, -87.7199738)  
4  09/16/2023 03:41:56 PM  41.796477 -87.708541  (41.796477414, -87.708540915)  

[5 rows x 22 columns]

In [58]:
df.shape

(632919, 22)

In [59]:
print('Original data has {} rows '.format(initial_df.shape[0]))
print('Updatated dataframe has {} row'.format(df.shape[0]))

Original data has 7898456 rows 
Updatated dataframe has 632919 row


## Irrelevant/Redundant Features <a id ='Irrelevant-/-Redundant-Features'></a> 

Upon initial review I noticed that some columns have irrelevant  data. Location, X coordinate and Y coordinate have the same information as Lon  and Lat columns. Also, Case Number columns have no significance information, since I  have ID case number. I will be dropping those columns.

In [60]:
df2 = df.drop(columns=['Case Number','X Coordinate', 'Y Coordinate','Location'])

In [61]:
# sanity check 
df2.head()

ID                    Date                 Block  IUCR  \
0  13203321  09/06/2023 05:00:00 PM      002XX N Wells st  1320   
1  13210088  08/31/2023 12:00:00 PM  023XX W JACKSON BLVD  1153   
2  13210004  07/24/2023 09:45:00 PM  073XX S JEFFERY BLVD  0281   
3  13210062  08/27/2023 07:00:00 AM  034XX N LAWNDALE AVE  0820   
4  13210107  09/04/2023 09:30:00 PM     053XX S HOMAN AVE  1310   

              Primary Type                          Description  \
0          CRIMINAL DAMAGE                           TO VEHICLE   
1       DECEPTIVE PRACTICE  FINANCIAL IDENTITY THEFT OVER $ 300   
2  CRIMINAL SEXUAL ASSAULT                       NON-AGGRAVATED   
3                    THEFT                       $500 AND UNDER   
4          CRIMINAL DAMAGE                          TO PROPERTY   

                     Location Description  Arrest  Domestic  Beat  District  \
0  PARKING LOT / GARAGE (NON RESIDENTIAL)   False     False   122       1.0   
1                                  STREET   False     False  1225      12.0   
2                               APARTMENT   False     False   333       3.0   
3                               APARTMENT   False     False  1732      17.0   
4                      RESIDENCE - GARAGE   False     False   822       8.0   

   Ward  Community Area FBI Code  Year              Updated On   Latitude  \
0  42.0            32.0       14  2023  09/14/2023 03:43:09 PM        NaN   
1  27.0            28.0       11  2023  09/16/2023 03:41:56 PM  41.877565   
2   7.0            43.0       02  2023  09/16/2023 03:41:56 PM  41.761919   
3  30.0            21.0       06  2023  09/16/2023 03:41:56 PM  41.943379   
4  14.0            63.0       14  2023  09/16/2023 03:41:56 PM  41.796477   

   Longitude  
0        NaN  
1 -87.684791  
2 -87.576209  
3 -87.719974  
4 -87.708541

In [62]:
df2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 632919 entries, 0 to 632918
Data columns (total 18 columns):
 #   Column                Non-Null Count   Dtype  
---  ------                --------------   -----  
 0   ID                    632919 non-null  int64  
 1   Date                  632919 non-null  object 
 2   Block                 632919 non-null  object 
 3   IUCR                  632919 non-null  object 
 4   Primary Type          632919 non-null  object 
 5   Description           632919 non-null  object 
 6   Location Description  630092 non-null  object 
 7   Arrest                632919 non-null  bool   
 8   Domestic              632919 non-null  bool   
 9   Beat                  632919 non-null  int64  
 10  District              632919 non-null  float64
 11  Ward                  632895 non-null  float64
 12  Community Area        632919 non-null  float64
 13  FBI Code              632919 non-null  object 
 14  Year                  632919 non-null  int64  
 15  

In [63]:
# Converting Date column and Updated On to a datetime format
df2['Date'] = pd.to_datetime(df2['Date'])
df2['Updated On'] = pd.to_datetime(df2['Updated On'])

In [64]:
# sanity check
df2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 632919 entries, 0 to 632918
Data columns (total 18 columns):
 #   Column                Non-Null Count   Dtype         
---  ------                --------------   -----         
 0   ID                    632919 non-null  int64         
 1   Date                  632919 non-null  datetime64[ns]
 2   Block                 632919 non-null  object        
 3   IUCR                  632919 non-null  object        
 4   Primary Type          632919 non-null  object        
 5   Description           632919 non-null  object        
 6   Location Description  630092 non-null  object        
 7   Arrest                632919 non-null  bool          
 8   Domestic              632919 non-null  bool          
 9   Beat                  632919 non-null  int64         
 10  District              632919 non-null  float64       
 11  Ward                  632895 non-null  float64       
 12  Community Area        632919 non-null  float64       
 13 

In [65]:
# Converting Arrest and Domestic columns to an int
df2['Arrest'] = df2['Arrest'].astype(int)
df2['Domestic'] = df2['Domestic'].astype(int)

In [66]:
#sanity check
df2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 632919 entries, 0 to 632918
Data columns (total 18 columns):
 #   Column                Non-Null Count   Dtype         
---  ------                --------------   -----         
 0   ID                    632919 non-null  int64         
 1   Date                  632919 non-null  datetime64[ns]
 2   Block                 632919 non-null  object        
 3   IUCR                  632919 non-null  object        
 4   Primary Type          632919 non-null  object        
 5   Description           632919 non-null  object        
 6   Location Description  630092 non-null  object        
 7   Arrest                632919 non-null  int32         
 8   Domestic              632919 non-null  int32         
 9   Beat                  632919 non-null  int64         
 10  District              632919 non-null  float64       
 11  Ward                  632895 non-null  float64       
 12  Community Area        632919 non-null  float64       
 13 

## Checking for duplicate data <a id ='Checking-for-Duplicate-Data'></a>

In [67]:
# checking for duplicated rows 
df2.duplicated().sum()

0

Based on the above results. There is no duplicated rows 

In [68]:
# checking for duplicated columns
df2.T.duplicated()

ID                      False
Date                    False
Block                   False
IUCR                    False
Primary Type            False
Description             False
Location Description    False
Arrest                  False
Domestic                False
Beat                    False
District                False
Ward                    False
Community Area          False
FBI Code                False
Year                    False
Updated On              False
Latitude                False
Longitude               False
dtype: bool

Creating a copy of the data frame to have a back up. Operation for duplicated columns takes a very long time to compute 

In [69]:
df3 = df2
# sanity check 
df3.head()

ID                Date                 Block  IUCR  \
0  13203321 2023-09-06 17:00:00      002XX N Wells st  1320   
1  13210088 2023-08-31 12:00:00  023XX W JACKSON BLVD  1153   
2  13210004 2023-07-24 21:45:00  073XX S JEFFERY BLVD  0281   
3  13210062 2023-08-27 07:00:00  034XX N LAWNDALE AVE  0820   
4  13210107 2023-09-04 21:30:00     053XX S HOMAN AVE  1310   

              Primary Type                          Description  \
0          CRIMINAL DAMAGE                           TO VEHICLE   
1       DECEPTIVE PRACTICE  FINANCIAL IDENTITY THEFT OVER $ 300   
2  CRIMINAL SEXUAL ASSAULT                       NON-AGGRAVATED   
3                    THEFT                       $500 AND UNDER   
4          CRIMINAL DAMAGE                          TO PROPERTY   

                     Location Description  Arrest  Domestic  Beat  District  \
0  PARKING LOT / GARAGE (NON RESIDENTIAL)       0         0   122       1.0   
1                                  STREET       0         0  1225      12.0   
2                               APARTMENT       0         0   333       3.0   
3                               APARTMENT       0         0  1732      17.0   
4                      RESIDENCE - GARAGE       0         0   822       8.0   

   Ward  Community Area FBI Code  Year          Updated On   Latitude  \
0  42.0            32.0       14  2023 2023-09-14 15:43:09        NaN   
1  27.0            28.0       11  2023 2023-09-16 15:41:56  41.877565   
2   7.0            43.0       02  2023 2023-09-16 15:41:56  41.761919   
3  30.0            21.0       06  2023 2023-09-16 15:41:56  41.943379   
4  14.0            63.0       14  2023 2023-09-16 15:41:56  41.796477   

   Longitude  
0        NaN  
1 -87.684791  
2 -87.576209  
3 -87.719974  
4 -87.708541

## Dealing with missing data <a id ='Dealing-with-Missing-Data'></a>

In [70]:
df3.isna().sum()

ID                          0
Date                        0
Block                       0
IUCR                        0
Primary Type                0
Description                 0
Location Description     2827
Arrest                      0
Domestic                    0
Beat                        0
District                    0
Ward                       24
Community Area              0
FBI Code                    0
Year                        0
Updated On                  0
Latitude                12828
Longitude               12828
dtype: int64

In [71]:
# Percentage of missing values in each column
df3.isna().sum() / df3.shape[0]*100

ID                      0.000000
Date                    0.000000
Block                   0.000000
IUCR                    0.000000
Primary Type            0.000000
Description             0.000000
Location Description    0.446661
Arrest                  0.000000
Domestic                0.000000
Beat                    0.000000
District                0.000000
Ward                    0.003792
Community Area          0.000000
FBI Code                0.000000
Year                    0.000000
Updated On              0.000000
Latitude                2.026800
Longitude               2.026800
dtype: float64

Four column have missing information:
-  Ward 0.038%
- Location Description 0.45%
- Latitude 2.03%
- Longitude 2.03%


In [72]:
#The Ward column has a very low percentage of missing information. Therefore, i will droop all rows where Ward is NAN
df3 = df3.dropna(subset=['Ward'])

In [73]:
# sanity check
df3.isna().sum()

ID                          0
Date                        0
Block                       0
IUCR                        0
Primary Type                0
Description                 0
Location Description     2826
Arrest                      0
Domestic                    0
Beat                        0
District                    0
Ward                        0
Community Area              0
FBI Code                    0
Year                        0
Updated On                  0
Latitude                12827
Longitude               12827
dtype: int64

In [74]:
# exploring row with NaN value for Location Description
nan_rows = df3[pd.isna(df3['Location Description'])]
nan_rows

ID                Date                  Block  IUCR  \
387     13207457 2023-09-06 14:20:00     032XX N OCONTO AVE  1153   
856     13209655 2023-05-30 09:30:00  003XX W WASHINGTON ST  1242   
1094    13203348 2023-09-07 14:35:00    052XX S NATCHEZ AVE  1153   
1471    13209657 2023-02-02 08:15:00      018XX N HONORE ST  1153   
1753    13211803 2023-08-31 11:14:00    020XX W DIVISION ST  1154   
...          ...                 ...                    ...   ...   
631781  13042362 2023-04-14 18:45:00       033XX W BYRON ST  1153   
632057  13078297 2023-02-01 12:00:00         031XX S MAY ST  1153   
632488  13048752 2023-04-19 22:35:00   112XX S EDBROOKE AVE  1153   
632655  13048696 2023-04-19 13:20:00     004XX N WABASH AVE  1153   
632816  13145202 2023-07-08 02:15:00   047XX W VAN BUREN ST  1153   

              Primary Type                              Description  \
387     DECEPTIVE PRACTICE      FINANCIAL IDENTITY THEFT OVER $ 300   
856     DECEPTIVE PRACTICE                           COMPUTER FRAUD   
1094    DECEPTIVE PRACTICE      FINANCIAL IDENTITY THEFT OVER $ 300   
1471    DECEPTIVE PRACTICE      FINANCIAL IDENTITY THEFT OVER $ 300   
1753    DECEPTIVE PRACTICE  FINANCIAL IDENTITY THEFT $300 AND UNDER   
...                    ...                                      ...   
631781  DECEPTIVE PRACTICE      FINANCIAL IDENTITY THEFT OVER $ 300   
632057  DECEPTIVE PRACTICE      FINANCIAL IDENTITY THEFT OVER $ 300   
632488  DECEPTIVE PRACTICE      FINANCIAL IDENTITY THEFT OVER $ 300   
632655  DECEPTIVE PRACTICE      FINANCIAL IDENTITY THEFT OVER $ 300   
632816  DECEPTIVE PRACTICE      FINANCIAL IDENTITY THEFT OVER $ 300   

       Location Description  Arrest  Domestic  Beat  District  Ward  \
387                     NaN       0         0  1631      16.0  29.0   
856                     NaN       0         0   122       1.0  42.0   
1094                    NaN       0         0   811       8.0  23.0   
1471                    NaN       0         0  1434      14.0  32.0   
1753                    NaN       0         0  1212      12.0   1.0   
...                     ...     ...       ...   ...       ...   ...   
631781                  NaN       0         0  1733      17.0  33.0   
632057                  NaN       0         0   913       9.0  11.0   
632488                  NaN       0         0   531       5.0   9.0   
632655                  NaN       0         0  1834      18.0  42.0   
632816                  NaN       0         0  1131      11.0  28.0   

        Community Area FBI Code  Year          Updated On   Latitude  \
387               17.0       11  2023 2023-09-14 15:43:09  41.938826   
856               32.0       11  2023 2023-09-16 15:41:56  41.883203   
1094              56.0       11  2023 2023-09-15 15:42:23  41.796829   
1471              22.0       11  2023 2023-09-15 15:41:25  41.914764   
1753              24.0       11  2023 2023-09-16 15:42:58        NaN   
...                ...      ...   ...                 ...        ...   
631781            16.0       11  2023 2023-08-19 15:40:26  41.952047   
632057            60.0       11  2023 2023-08-19 15:40:26  41.836678   
632488            49.0       11  2023 2023-08-19 15:40:26  41.690268   
632655             8.0       11  2023 2023-08-19 15:40:26  41.888994   
632816            25.0       11  2023 2023-08-19 15:40:26  41.875039   

        Longitude  
387    -87.807986  
856    -87.635823  
1094   -87.784389  
1471   -87.673903  
1753          NaN  
...           ...  
631781 -87.712185  
632057 -87.654768  
632488 -87.619568  
632655 -87.626935  
632816 -87.743643  

[2826 rows x 18 columns]

In [75]:
# Filling NaN value for Location Description with 'Unknown'
df3['Location Description'] = df3['Location Description'].fillna('Unknown')

C:\Users\vikst\AppData\Local\Temp\ipykernel_19804\1054172993.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df3['Location Description'] = df3['Location Description'].fillna('Unknown')


In [76]:
# sanity check 
df3.isna().sum()

ID                          0
Date                        0
Block                       0
IUCR                        0
Primary Type                0
Description                 0
Location Description        0
Arrest                      0
Domestic                    0
Beat                        0
District                    0
Ward                        0
Community Area              0
FBI Code                    0
Year                        0
Updated On                  0
Latitude                12827
Longitude               12827
dtype: int64

In [77]:
# getting the number of rows where the values for Latitude and Longitude are NAN
rows_with_nan_values = df3[df3['Latitude'].isna() & df3['Longitude'].isna()]           

In [78]:
# sanity check
rows_with_nan_values

ID                Date                    Block  IUCR  \
0       13203321 2023-09-06 17:00:00         002XX N Wells st  1320   
18      12955336 2022-11-16 10:23:00     087XX S KINGSTON AVE  1562   
19      12978205 2022-08-01 22:00:00       028XX S KEELER AVE  1562   
20      13041906 2022-10-15 01:30:00   051XX W WRIGHTWOOD AVE  0486   
147     13209581 2021-08-01 00:00:00          012XX E 78TH ST  1563   
...          ...                 ...                      ...   ...   
632192  13062643 2022-04-12 00:00:00       074XX S WABASH AVE  1154   
632242  13092803 2021-08-23 00:00:00  036XX W SHAKESPEARE AVE  1563   
632471  13099981 2022-03-01 00:00:00       019XX S WABASH AVE  0460   
632549  13168008 2023-08-05 02:30:00     017XX W Van Buren St  0810   
632668  13138379 2023-06-28 21:30:00      051XX S Cornell Ave  0810   

              Primary Type                              Description  \
0          CRIMINAL DAMAGE                               TO VEHICLE   
18             SEX OFFENSE         AGGRAVATED CRIMINAL SEXUAL ABUSE   
19             SEX OFFENSE         AGGRAVATED CRIMINAL SEXUAL ABUSE   
20                 BATTERY                  DOMESTIC BATTERY SIMPLE   
147            SEX OFFENSE                    CRIMINAL SEXUAL ABUSE   
...                    ...                                      ...   
632192  DECEPTIVE PRACTICE  FINANCIAL IDENTITY THEFT $300 AND UNDER   
632242         SEX OFFENSE                    CRIMINAL SEXUAL ABUSE   
632471             BATTERY                                   SIMPLE   
632549               THEFT                                OVER $500   
632668               THEFT                                OVER $500   

                          Location Description  Arrest  Domestic  Beat  \
0       PARKING LOT / GARAGE (NON RESIDENTIAL)       0         0   122   
18                                   RESIDENCE       0         1   423   
19                                   APARTMENT       0         1  1031   
20                                   APARTMENT       1         1  2521   
147                                  APARTMENT       0         0   411   
...                                        ...     ...       ...   ...   
632192                               RESIDENCE       0         0   323   
632242                               RESIDENCE       0         1  2525   
632471                               APARTMENT       0         0   131   
632549             HOSPITAL BUILDING / GROUNDS       0         0  1231   
632668                                  STREET       0         0   234   

        District  Ward  Community Area FBI Code  Year          Updated On  \
0            1.0  42.0            32.0       14  2023 2023-09-14 15:43:09   
18           4.0   7.0            46.0       17  2022 2023-09-14 15:41:59   
19          10.0  22.0            30.0       17  2022 2023-09-14 15:41:59   
20          25.0  31.0            19.0      08B  2022 2023-09-14 15:41:59   
147          4.0   8.0            45.0       17  2021 2023-09-14 15:43:09   
...          ...   ...             ...      ...   ...                 ...   
632192       3.0   6.0            69.0       11  2022 2023-05-05 15:41:14   
632242      25.0  35.0            22.0       17  2021 2023-06-02 15:41:42   
632471       1.0   3.0            33.0      08B  2022 2023-06-08 15:42:44   
632549      12.0  27.0            28.0       06  2023 2023-08-12 15:41:48   
632668       2.0   5.0            41.0       06  2023 2023-07-13 15:43:14   

        Latitude  Longitude  
0            NaN        NaN  
18           NaN        NaN  
19           NaN        NaN  
20           NaN        NaN  
147          NaN        NaN  
...          ...        ...  
632192       NaN        NaN  
632242       NaN        NaN  
632471       NaN        NaN  
632549       NaN        NaN  
632668       NaN        NaN  

[12827 rows x 18 columns]

In [79]:
# Getting unigue block names 
unique_blocks = rows_with_nan_values['Block'].value_counts()

print(unique_blocks)

033XX W FILLMORE ST         428
006XX W OHARE ST            136
100XX W OHARE ST             71
014XX S MUSEUM CAMPUS DR     70
003XX E RANDOLPH ST          62
                           ... 
017XX W 48TH ST               1
088XX S STONY ISLAND AVE      1
002XX W 31ST ST               1
035XX W SCHRAEDER DR          1
051XX S Cornell Ave           1
Name: Block, Length: 7571, dtype: int64


I encountered a situation where 12,827 rows lacked longitude and latitude details for various city blocks, out of which 7,571 were distinct blocks in Chicago. Although there were methods like calculating mean values or using interpolation to fill these gaps, these approaches were deemed inaccurate. Considering only 2% of the rows were affected, I opted to remove the rows with missing data

In [80]:
# droping colums where Latitude and Longitude have NAN values
df3 = df3.dropna(subset=['Latitude','Longitude'])

In [81]:
#sanity check 
df3.isna().sum()

ID                      0
Date                    0
Block                   0
IUCR                    0
Primary Type            0
Description             0
Location Description    0
Arrest                  0
Domestic                0
Beat                    0
District                0
Ward                    0
Community Area          0
FBI Code                0
Year                    0
Updated On              0
Latitude                0
Longitude               0
dtype: int64

## Saving the data <a id='Saving-the-Data'></a>

In [82]:
# Saving the cleaned data for future analysis 
df3.to_csv('/Users/vikst/Desktop/data/Chicago_Crime-cleaned.csv', index=False)